packages/squiggle-lang/src/rescript/GenericDist/GenericDist_GenericOperation.res

```diff
@@ -0,0 +1,151 @@
+type operation = GenericDist_Types.Operation.genericFunctionCallInfo
+type genericDist = GenericDist_Types.genericDist
+type error = GenericDist_Types.error
+
+// TODO: It could be great to use a cache for some calculations (basically, do memoization). Also, better analytics/tracking could go a long way.
+
+type params = {
+  sampleCount: int,
+  xyPointLength: int,
+}
+
+type outputType = [
+  | #Dist(genericDist)
+  | #Float(float)
+  | #String(string)
+  | #GenDistError(error)
+]
+
+module Output = {
+  let toDist = (o: outputType) =>
+    switch o {
+    | #Dist(d) => Some(d)
+    | _ => None
+    }
+
+  let toFloat = (o: outputType) =>
+    switch o {
+    | #Float(d) => Some(d)
+    | _ => None
+    }
+
+  let toString = (o: outputType) =>
+    switch o {
+    | #String(d) => Some(d)
+    | _ => None
+    }
+
+  let toError = (o: outputType) =>
+    switch o {
+    | #GenDistError(d) => Some(d)
+    | _ => None
+    }
+}
+
+let fromResult = (r: result<outputType, error>): outputType =>
+  switch r {
+  | Ok(o) => o
+  | Error(e) => #GenDistError(e)
+  }
+
+let outputToDistResult = (b: outputType): result<genericDist, error> =>
+  switch b {
+  | #Dist(r) => Ok(r)
+  | #GenDistError(r) => Error(r)
+  | _ => Error(Unreachable)
+  }
+
+let rec run = (extra, fnName: operation): outputType => {
+  let {sampleCount, xyPointLength} = extra
+
+  let reCall = (~extra=extra, ~fnName=fnName, ()) => {
+    run(extra, fnName)
+  }
+
+  let toPointSet = r => {
+    switch reCall(~fnName=#fromDist(#toDist(#toPointSet), r), ()) {
+    | #Dist(#PointSet(p)) => Ok(p)
+    | #GenDistError(r) => Error(r)
+    | _ => Error(Unreachable)
+    }
+  }
+
+  let toSampleSet = r => {
+    switch reCall(~fnName=#fromDist(#toDist(#toSampleSet(sampleCount)), r), ()) {
+    | #Dist(#SampleSet(p)) => Ok(p)
+    | #GenDistError(r) => Error(r)
+    | _ => Error(Unreachable)
+    }
+  }
+
+  let scaleMultiply = (r, weight) =>
+    reCall(
+      ~fnName=#fromDist(#toDistCombination(#Pointwise, #Multiply, #Float(weight)), r),
+      (),
+    )->outputToDistResult
+
+  let pointwiseAdd = (r1, r2) =>
+    reCall(
+      ~fnName=#fromDist(#toDistCombination(#Pointwise, #Add, #Dist(r2)), r1),
+      (),
+    )->outputToDistResult
+
+  let fromDistFn = (subFn: GenericDist_Types.Operation.fromDist, dist: genericDist) =>
```

Comment:
Is this like, part of/a utility for an AST parser? I might be wrong, but something makes me feel like the AST parsing should be part of an AST parsing module, and not part of a distribution library?

Reply:
Yea; the GenericOperation module is basically a tiny AST, with just function calls. I think it's good to have isolated because it's pretty specialized. This helps with testing and modularity. 

It's very possible there's some other clean way to modularize this sort of thing.